### PR TITLE
External CI: enable ROCR-Runtime tests

### PR DESCRIPTION
**Pull request by danielsu-amd**
Builds and runs kfdtest, rdmatest, and rocrtst. rdmatest does not publish test results.

rocrtst subtest `rocrtstNeg.Memory_Negative_Tests` is currently failing and putting GPUs in a bad state. An amdgpu patch for it is targetted for a later ROCm release, so that is being filtered out for now.

Successful run: https://dev.azure.com/ROCm-CI/ROCm-CI/_build/results?buildId=8753&view=results

### CODE REVIEW

File: .azuredevops/components/ROCR-Runtime.yml [hunk 1]
```diff
@@ -50,3 +56,123 @@ jobs:
...
+  - template: ${{ variables.CI_TEMPLATE_PATH }}/steps/dependencies-rocm.yml
+    parameters:
+      dependencyList: ${{ parameters.rocmTestDependencies }}
+      gpuTarget: $(JOB_GPU_TARGET)
+      ${{ if eq(parameters.checkoutRef, '') }}:
+        dependencySource: staging
+      ${{ elseif ne(parameters.checkoutRef, '') }}:
+        dependencySource: tag-builds
+  - task: Bash@3
+    displayName: Clone ROCR-Runtime
```
Comment (amd-jmacaran) [inline .azuredevops/components/ROCR-Runtime.yml:103]: Change this to an Azure checkout so it doesn't just clone default branch and clones the commit/PR in question to account for test changes in those commits/PRs.

Comment (danielsu-amd) [reply .azuredevops/components/ROCR-Runtime.yml]: Changed it to use the checkout template

File: .azuredevops/components/ROCR-Runtime.yml [hunk 2]
```diff
@@ -50,3 +56,120 @@ jobs:
...
+      script: |
+        mkdir build && cd build
+        cmake -DCMAKE_PREFIX_PATH=$(Agent.BuildDirectory)/rocm ..
+        make
+  - template: ${{ variables.CI_TEMPLATE_PATH }}/steps/test.yml
+    parameters:
+      componentName: kfdtest
+      testExecutable: BIN_DIR=$(Agent.BuildDirectory)/ROCR-Runtime/libhsakmt/tests/kfdtest/build ./run_kfdtest.sh
+      testParameters: '-p core --gtest_output=xml:./test_output.xml --gtest_color=yes'
+      testDir: $(Agent.BuildDirectory)/ROCR-Runtime/libhsakmt/tests/kfdtest/scripts
```
Comment (amd-jmacaran) [inline .azuredevops/components/ROCR-Runtime.yml:119]: Change $(Agent.BuildDirectory)/ROCR-Runtime to $(Build.SourcesDirectory) to account for checkout step usage.

File: .azuredevops/components/ROCR-Runtime.yml [hunk 3]
```diff
@@ -50,3 +56,120 @@ jobs:
...
+      targetType: 'inline'
+      workingDirectory: $(Agent.BuildDirectory)/ROCR-Runtime/libhsakmt/tests/rdma/simple/app
+      script: |
+        cmake -DCMAKE_PREFIX_PATH=$(Agent.BuildDirectory)/rocm -DLIBHSAKMT_PATH=$(Agent.BuildDirectory)/rocm -DDRM_AMDGPU_INCLUDE_DIRS=$(Agent.BuildDirectory)/rocm/include .
+        cmake --build .
+  - task: Bash@3
+    displayName: Build rdmatest driver
+    inputs:
+      targetType: 'inline'
+      workingDirectory: $(Agent.BuildDirectory)/ROCR-Runtime/libhsakmt/tests/rdma/simple/drv
```
Comment (amd-jmacaran) [inline .azuredevops/components/ROCR-Runtime.yml:132]: Similar as above

File: .azuredevops/components/ROCR-Runtime.yml [hunk 4]
```diff
@@ -50,3 +56,120 @@ jobs:
...
+    parameters:
+      componentName: kfdtest
+      testExecutable: BIN_DIR=$(Agent.BuildDirectory)/ROCR-Runtime/libhsakmt/tests/kfdtest/build ./run_kfdtest.sh
+      testParameters: '-p core --gtest_output=xml:./test_output.xml --gtest_color=yes'
+      testDir: $(Agent.BuildDirectory)/ROCR-Runtime/libhsakmt/tests/kfdtest/scripts
+  - task: Bash@3
+    displayName: Build rdmatest app
+    inputs:
+      targetType: 'inline'
+      workingDirectory: $(Agent.BuildDirectory)/ROCR-Runtime/libhsakmt/tests/rdma/simple/app
```
Comment (amd-jmacaran) [inline .azuredevops/components/ROCR-Runtime.yml:124]: Similar as above

File: .azuredevops/components/ROCR-Runtime.yml [hunk 5]
```diff
@@ -50,3 +56,120 @@ jobs:
...
+      ${{ elseif ne(parameters.checkoutRef, '') }}:
+        dependencySource: tag-builds
+  - template: ${{ variables.CI_TEMPLATE_PATH }}/steps/checkout.yml
+    parameters:
+      checkoutRepo: ${{ parameters.checkoutRepo }}
+  - task: Bash@3
+    displayName: Build kfdtest
+    inputs:
+      targetType: 'inline'
+      workingDirectory: $(Agent.BuildDirectory)/ROCR-Runtime/libhsakmt/tests/kfdtest
```
Comment (amd-jmacaran) [inline .azuredevops/components/ROCR-Runtime.yml:109]: Refer to below

File: .azuredevops/components/ROCR-Runtime.yml [hunk 6]
```diff
@@ -50,3 +56,120 @@ jobs:
...
+      workingDirectory: $(Agent.BuildDirectory)/ROCR-Runtime/libhsakmt/tests/rdma/simple/drv
+      script: |
+        sed -i 's/HSAKMT_PAGE_SHIFT/PAGE_SHIFT/g' amdp2ptest.c
+        sed -i 's/"MIT"/"GPL"/' amdp2ptest.c
+        RDMA_HEADER_DIR=/usr/src/amdgpu-*/include make all
+  - task: Bash@3
+    displayName: Install rdmatest driver
+    inputs:
+      targetType: 'inline'
+      workingDirectory: $(Agent.BuildDirectory)/ROCR-Runtime/libhsakmt/tests/rdma/simple/drv
```
Comment (amd-jmacaran) [inline .azuredevops/components/ROCR-Runtime.yml:141]: Similar as above

File: .azuredevops/components/ROCR-Runtime.yml [hunk 7]
```diff
@@ -50,3 +56,120 @@ jobs:
...
+      script: |
+        sudo rmmod amdp2ptest.ko
+        sudo insmod amdp2ptest.ko
+  - template: ${{ variables.CI_TEMPLATE_PATH }}/steps/test.yml
+    parameters:
+      componentName: rdmatest
+      testExecutable: yes | ./rdma_test
+      testParameters: ''
+      testDir: $(Agent.BuildDirectory)/ROCR-Runtime/libhsakmt/tests/rdma/simple/app
```
Comment (amd-jmacaran) [inline .azuredevops/components/ROCR-Runtime.yml:150]: Similar as above

File: .azuredevops/components/ROCR-Runtime.yml [hunk 8]
```diff
@@ -50,3 +56,120 @@ jobs:
...
+      testPublishResults: false
+  - task: Bash@3
+    displayName: Build rocrtst
+    inputs:
+      targetType: 'inline'
+      workingDirectory: $(Agent.BuildDirectory)/ROCR-Runtime/rocrtst/suites/test_common
```
Comment (amd-jmacaran) [inline .azuredevops/components/ROCR-Runtime.yml:156]: Similar as above

File: .azuredevops/components/ROCR-Runtime.yml [hunk 9]
```diff
@@ -50,3 +56,120 @@ jobs:
...
+      script: |
+        sudo rm -rf $(Agent.BuildDirectory)/external/llvm-project
```
Comment (amd-jmacaran) [inline .azuredevops/components/ROCR-Runtime.yml:158]: Similar as above

File: .azuredevops/components/ROCR-Runtime.yml [hunk 10]
```diff
@@ -50,3 +56,120 @@ jobs:
...
+          -DLLVM_DIR=$(Agent.BuildDirectory)/rocm/llvm/bin \
+          -DOPENCL_DIR=$(Agent.BuildDirectory)/rocm/llvm/bin
+        make
+        make rocrtst_kernels
+  - template: ${{ variables.CI_TEMPLATE_PATH }}/steps/test.yml
+    parameters:
+      componentName: rocrtst
+      testExecutable: ./rocrtst64
+      testParameters: '--gtest_filter="-rocrtstNeg.Memory_Negative_Tests" --gtest_output=xml:./test_output.xml --gtest_color=yes'
+      testDir: $(Agent.BuildDirectory)/ROCR-Runtime/rocrtst/suites/test_common/build/$(JOB_GPU_TARGET)
```
Comment (amd-jmacaran) [inline .azuredevops/components/ROCR-Runtime.yml:175]: Similar as above

File: .azuredevops/components/ROCR-Runtime.yml (hunk 9 above)
Comment (danielsu-amd) [reply .azuredevops/components/ROCR-Runtime.yml:158]: This should stay as is, the symlink is for rocrtst to find openCL files which it looks for in
`${OPENCL_DIR}/../../../external/llvm-project/clang/lib/Headers/`
which is
`$(Agent.BuildDirectory)/external/llvm-project/clang/lib/Headers/`

https://github.com/ROCm/ROCR-Runtime/blob/a926a070ee9f18a1043a33df0fe1f33bcd7c616f/rocrtst/suites/test_common/CMakeLists.txt#L158

**Comment by danielsu-amd** [comment]
New build log after using checkout template and switching paths to Build.SourcesDirectory:
https://dev.azure.com/ROCm-CI/ROCm-CI/_build/results?buildId=8840&view=results

**Pull request merged**